pkg/identity/gitlabcom/principal.go

```diff
@@ -0,0 +1,205 @@
+		PipelineSource    string `json:"pipeline_source"`
+		PipelineID        string `json:"pipeline_id"`
+		NamespacePath     string `json:"namespace_path"`
+		NamespaceID       string `json:"namespace_id"`
+		JobID             string `json:"job_id"`
+		Ref               string `json:"ref"`
+		RefType           string `json:"ref_type"`
+		Sha               string `json:"sha"`
+		RunnerEnvironment string `json:"runner_environment"`
+		RunnerID          int64  `json:"runner_id"`
+	}
+
+	if err := token.Claims(&claims); err != nil {
+		return nil, err
+	}
+
+	if claims.ProjectPath == "" {
+		return nil, errors.New("missing project_path claim in ID token")
+	}
+
+	if claims.PipelineSource == "" {
+		return nil, errors.New("missing pipeline_source claim in ID token")
+	}
+
+	if claims.PipelineID == "" {
+		return nil, errors.New("missing pipeline_id claim in ID token")
+	}
+
+	if claims.JobID == "" {
+		return nil, errors.New("missing job_id claim in ID token")
+	}
+
+	if claims.Ref == "" {
+		return nil, errors.New("missing ref claim in ID token")
+	}
+
+	if claims.RefType == "" {
+		return nil, errors.New("missing ref_type claim in ID token")
+	}
+
+	if claims.NamespacePath == "" {
+		return nil, errors.New("missing namespace_path claim in ID token")
+	}
+
+	if claims.NamespaceID == "" {
+		return nil, errors.New("missing namespace_id claim in ID token")
+	}
+
+	if claims.ProjectID == "" {
+		return nil, errors.New("missing project_id claim in ID token")
+	}
+
+	if claims.Sha == "" {
+		return nil, errors.New("missing sha claim in ID token")
+	}
+
+	if claims.RunnerEnvironment == "" {
+		return nil, errors.New("missing runner_environment claim in ID token")
+	}
+
+	if claims.RunnerID == 0 {
+		return nil, errors.New("missing runner_id claim in ID token")
+	}
+
+	var ref string
+	switch claims.RefType {
+	case "branch":
+		ref = "refs/heads/" + claims.Ref
+	case "tag":
+		ref = "refs/tags/" + claims.Ref
+	default:
+		return nil, fmt.Errorf("unexpected ref_type: %s", claims.RefType)
+	}
+
+	return &jobPrincipal{
+		subject:           token.Subject,
+		issuer:            token.Issuer,
+		url:               `https://gitlab.com/`,
+		eventName:         claims.PipelineSource,
+		pipelineID:        claims.PipelineID,
+		repository:        claims.ProjectPath,
+		ref:               ref,
+		repositoryID:      claims.ProjectID,
+		repositoryOwner:   claims.NamespacePath,
+		repositoryOwnerID: claims.NamespaceID,
+		jobID:             claims.JobID,
+		sha:               claims.Sha,
+		runnerID:          claims.RunnerID,
+		runnerEnvironment: claims.RunnerEnvironment,
+	}, nil
+}
+
+func (p jobPrincipal) Name(_ context.Context) string {
+	return p.subject
+}
+
+func (p jobPrincipal) Embed(_ context.Context, cert *x509.Certificate) error {
+	baseURL, err := url.Parse(p.url)
+	if err != nil {
+		return err
+	}
+
+	// Set workflow ref URL to SubjectAlternativeName on certificate
+	cert.URIs = []*url.URL{baseURL.JoinPath(p.repository, "/-/jobs/", p.jobID)}
```

Comment:
@wlynch Is this correct for the SAN? I recall you said that job was too granular. pipeline_ref? Or something else?

Reply:
Good catch - you're right. This should be 
```suggestion
	cert.URIs = []*url.URL{baseURL.JoinPath(p.repository, "@", ref)}
```

Reply:
Is pipeline ref too granular? That represents the build configuration, but I could imagine creating policy around that. 

If not, for ref, i think we’d need a different delimiter, unless @ creates a valid uri

Reply:
LGTM on @ in the SAN.

Hm, I definitely agree that if pipeline_ref points to a URI outside of gitlab, then we should avoid that. I'm good with this then. Thanks for the background!